kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKms.java

```diff
@@ -0,0 +1,173 @@
+
+    }
+
+    /**
+     * {@inheritDoc}
+     * <br/>
+     * @see <a href="https://developer.hashicorp.com/vault/api-docs/secret/transit#decrypt">https://developer.hashicorp.com/vault/api-docs/secret/transit#decrypt</a>
+     */
+    @NonNull
+    @Override
+    public CompletionStage<SecretKey> decryptEdek(@NonNull VaultEdek edek) {
+
+        var body = createDecryptPostBody(edek);
+
+        var request = createVaultRequest()
+                .uri(vaultUrl.resolve("v1/transit/decrypt/%s".formatted(edek.kekRef())))
+                .POST(HttpRequest.BodyPublishers.ofString(body))
+                .build();
+
+        return vaultClient.sendAsync(request, statusHandler(edek.kekRef(), new JsonBodyHandler<VaultResponse<DecryptData>>(new TypeReference<>() {
+        }))).thenApply(HttpResponse::body)
+                .thenApply(Supplier::get)
+                .thenApply(VaultResponse::data)
+                .thenApply(data -> new SecretKeySpec(Base64.getDecoder().decode(data.plaintext()), AES_KEY_ALGO));
+    }
+
+    private String createDecryptPostBody(@NonNull VaultEdek edek) {
+        var map = Map.of("ciphertext", new String(edek.edek(), StandardCharsets.UTF_8));
+
+        try {
+            return new ObjectMapper().writeValueAsString(map);
+        }
+        catch (JsonProcessingException e) {
+            throw new KmsException("Failed to build request body for %s".formatted(edek.kekRef()));
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * <br/>
+     * @see <a href="https://developer.hashicorp.com/vault/api-docs/secret/transit#read-key">https://developer.hashicorp.com/vault/api-docs/secret/transit#read-key</a>
+     */
+    @NonNull
+    @Override
+    public CompletableFuture<String> resolveAlias(@NonNull String alias) {
+
+        var request = createVaultRequest()
+                .uri(vaultUrl.resolve("v1/transit/keys/%s".formatted(alias)))
+                .build();
+
+        return vaultClient.sendAsync(request, statusHandler(alias, new JsonBodyHandler<VaultResponse<ReadKeyData>>(new TypeReference<>() {
+        })))
+                .thenApply(HttpResponse::body)
+                .thenApply(Supplier::get)
+                .thenApply(VaultResponse::data)
+                .thenApply(ReadKeyData::name);
+
+    }
+
+    @NonNull
+    @Override
+    public Serde<VaultEdek> edekSerde() {
+        return new VaultEdekSerde();
+    }
+
+    private HttpRequest.Builder createVaultRequest() {
+        return HttpRequest.newBuilder()
+                .header("X-Vault-Token", vaultToken)
+                .header("Accept", "application/json");
+    }
+
+    private static <T> HttpResponse.BodyHandler<T> statusHandler(String keyRef, HttpResponse.BodyHandler<T> handler) {
+        return r -> {
+            if (r.statusCode() == 404 || r.statusCode() == 400) {
+                throw new UnknownKeyException("key '%s' is not found.".formatted(keyRef));
+            }
+            else if (r.statusCode() != 200) {
```

Comment:
Do we need some special handling around `204` success with no data returned?

I suspect a lot of other bits of the vault KMS would be unhappy with a 204 response but it is supposed to be a success...

Reply:
I believe all our API interactions with Vault require a response, so I think the code is right to expect only 200.